doc/developer/design/20250415_large_select_result_size.md

```diff
@@ -0,0 +1,174 @@
+
+1. Fast-path SELECT: there is an existing arrangement (index) that we can
+   extract our result from.
+2. Slow-path SELECT: in order to get our result we first need to render a
+   temporary dataflow with an index that will contain the results. Then we can
+   use the same logic as a fast-path SELECT to extract the result. Afterwards
+   the temporary dataflow is torn down again.
+3. Fast-path persist SELECT: The query is simple enough that we can directly
+   extract it from a persist shard. With potentially some amount of
+   map-filter-project applied. This happens on the `clusterd` side and is
+   shipped back using the same mechanism that is used for shipping back other
+   SELECT results.
+
+1 and 2 here imply that the result must fit into memory on the cluster side,
+unless we want to, say, change how results can be extracted on the cluster
+side.
+
+## Goals
+
+The user facing goal is:
+
+- Lift result-size limitation for streamable queries
+
+And we have technical goals as well, enabling potential future work and making
+sure we have the right abstractions in place:
+
+- Create re-usable building blocks for shipping data/results out-of-band, these
+  can then be used in other places, such as for SUBSCRIBE
+- Make persist Batches more of a "first-class" citizen, eventually allowing us
+  to decouple where and how batches are written from where and how they are
+  appended to a shard
+
+## Non-Goals
+
+- Lift limitation on result having to fit into cluster memory: for non-persist
+  queries, the result is still first staged in an arrangement before we read it
+  out and send it to `environmentd`. Or we read out of an existing arrangement
+- Change how results are extracted from "the dataflows" on the cluster side
+- Lift result-size limitation for non-streamable queries
+- Change result-size limitations for SUBSCRIBE
+- Change how we ship around and append data to shards, when handling INSERT
+- Change how we ship around and append data for sources
+
+## Implementation
+
+Purposefully high level, but the rough idea is that we a) need to stop sending
+large amounts of data through the compute protocol, and b) need to stop
+materializing large results in `environmentd` memory.
+
+We want to re-use the existing abstractions we have around the persist blob
+store and batches. We use persist batches to "stash" data that we can then read
+out again in other places, without intending to append these batches to a
+persist shard. Though the latter is desirable in future work of decoupling
+creating and appending batches.
+
+For our user-facing goal specifically: on the cluster side, we "stash" query
+results as persist batches, send a handle to that data back via the compute
+protocol, and then (in `environmentd`) stream those results from the persist
+blob/batch store back to the client.
+
+We would use a (configurable) threshold for determining whether to send a
+result inline in the compute protocol or out-of-band via the blob store.
```

Comment:
Related: moving the task of writing data to envd would make it simple to determine whether a result is small or large.